# Data management using Pandas

**Data management** is a crucial component to statistical analysis and data science work.

This notebook will show you how to import, view, undertand, and manage your data using the [Pandas](http://pandas.pydata.org) data processing library, i.e., the notebook will demonstrates how to read a dataset into Python, and obtain a basic understanding of its content.

Note that **Python** by itself is a general-purpose programming language and does not provide high-level data processing capabilities.  The **Pandas** library was developed to meet this need. **Pandas** is the most popular Python library for data manipulation, and we will use it extensively in this course. **Pandas** provides high-performance, easy-to-use data structures and data analysis tools.

The main data structure that **Pandas** works with is called a **Data Frame**. This is a two-dimensional table of data in which the rows typically
represent cases and the columns represent variables (e.g. data used in this tutorial).  Pandas also has a one-dimensional data structure called a **Series** that we will encounter when accesing a single column of a Data Frame.

Pandas has a variety of functions named `read_xxx` for reading data in different formats.  Right now we will focus on reading `csv` files, which stands for comma-separated values. However the other file formats include `excel`, `json`, and `sql`.

There are many other options to `read_csv` that are very useful.  For example, you would use the option `sep='\t'` instead of the default `sep=','` if the fields of your data file are delimited by tabs instead of commas.  See [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) for the full documentation for `read_csv`.


## Acknowledgments

- The dataset used in this tutorial is from https://www.coursera.org/ from the course "Understanding and Visualizing Data with Python" by University of Michigan


# Importing libraries


In [1]:
# Import the packages that we will be using
import pandas as pd                 # For data handling

#pd.set_option('display.max_columns', 100) # Show all columns when looking at dataframe

# Importing data

In [2]:
# Define where you are running the code: colab or local
RunInColab          = True     # (False: no  | True: yes)

# If running in colab:
if RunInColab:
    # Mount your google drive in google colab
    from google.colab import drive
    drive.mount('/content/drive')

    # Find location
    #!pwd
    #!ls
    #!ls "/content/drive/My Drive/Colab Notebooks/MachineLearningWithPython/"

    # Define path del proyecto
    Ruta            = "/content/drive/MyDrive/A01642269/TC1002S/"

else:
    # Define path del proyecto
    Ruta            = ""

Mounted at /content/drive


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# url string that hosts our .csv file
url = Ruta + "datasets/iris/iris.csv"

# Read the .csv file and store it as a pandas Data Frame
df = pd.read_csv(url)



If we want to print the information about th output object type we would simply type the following:

In [5]:
type(df)

pandas.core.frame.DataFrame

# Exploring the content of the data set

Use the `shape` method to determine the numbers of rows and columns in a data frame. This can be used to confirm that we have actually obtained the data the we are expecting.

Based on what we see below, the data set being read here has $N_r$ rows, corresponding to $N_r$ observations, and $N_c$ columns, corresponding to $N_c$ variables in this particular data file.

In [6]:
df.shape

(150, 5)

In [7]:
# Print the number of rows
Nrows = df.shape[0]
Nrows

150

In [8]:
# Print the number of columns
Ncols = df.shape[1]
Ncols

5

If we want to show the entire data frame we would simply write the following:

In [9]:
df

Petal_x  Petal_y  Sepal_x  Sepal_y            Type
0        5.1      3.5      1.4      0.2     Iris-setosa
1        4.9      3.0      1.4      0.2     Iris-setosa
2        4.7      3.2      1.3      0.2     Iris-setosa
3        4.6      3.1      1.5      0.2     Iris-setosa
4        5.0      3.6      1.4      0.2     Iris-setosa
..       ...      ...      ...      ...             ...
145      6.7      3.0      5.2      2.3  Iris-virginica
146      6.3      2.5      5.0      1.9  Iris-virginica
147      6.5      3.0      5.2      2.0  Iris-virginica
148      6.2      3.4      5.4      2.3  Iris-virginica
149      5.9      3.0      5.1      1.8  Iris-virginica

[150 rows x 5 columns]

As you can see, we have a 2-Dimensional object where each row is an independent observation and each coloum is a variable.

Now, use the the `head()` function to show the first 5 rows of our data frame

In [10]:
df.head()
df.head(10)

Petal_x  Petal_y  Sepal_x  Sepal_y         Type
0      5.1      3.5      1.4      0.2  Iris-setosa
1      4.9      3.0      1.4      0.2  Iris-setosa
2      4.7      3.2      1.3      0.2  Iris-setosa
3      4.6      3.1      1.5      0.2  Iris-setosa
4      5.0      3.6      1.4      0.2  Iris-setosa
5      5.4      3.9      1.7      0.4  Iris-setosa
6      4.6      3.4      1.4      0.3  Iris-setosa
7      5.0      3.4      1.5      0.2  Iris-setosa
8      4.4      2.9      1.4      0.2  Iris-setosa
9      4.9      3.1      1.5      0.1  Iris-setosa

Also, you can use the the `tail()` function to show the last 5 rows of our data frame

In [11]:
df.tail()
df.tail(3)

Petal_x  Petal_y  Sepal_x  Sepal_y            Type
147      6.5      3.0      5.2      2.0  Iris-virginica
148      6.2      3.4      5.4      2.3  Iris-virginica
149      5.9      3.0      5.1      1.8  Iris-virginica

The columns in a Pandas data frame have names, to see the names, use the `columns` method:

To gather more information regarding the data, we can view the column names with the following function:

In [12]:
df.columns

Index(['Petal_x', 'Petal_y', 'Sepal_x', 'Sepal_y', 'Type'], dtype='object')

Be aware that every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the `dtypes` attribute of the data frame.

In [13]:
df.dtypes

Petal_x    float64
Petal_y    float64
Sepal_x    float64
Sepal_y    float64
Type        object
dtype: object

Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important.

In [14]:
# Summary statistics for the quantitative variables
df.describe()


Petal_x     Petal_y     Sepal_x     Sepal_y
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.057333    3.758000    1.199333
std      0.828066    0.435866    1.765298    0.762238
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [15]:
# Drop observations with NaN values
df.Petal_x.dropna().describe()
#df.Wingspan.dropna().describe()

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: Petal_x, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows

- `df.mean()` Returns the mean of all columns
- `df.corr()` Returns the correlation between columns in a data frame
- `df.count()` Returns the number of non-null values in each data frame column
- `df.max()` Returns the highest value in each column
- `df.min()` Returns the lowest value in each column
- `df.median()` Returns the median of each column
- `df.std()` Returns the standard deviation of each column

In [16]:
df.mean()

<ipython-input-16-c61f0c8f89b5>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()


Petal_x    5.843333
Petal_y    3.057333
Sepal_x    3.758000
Sepal_y    1.199333
dtype: float64

# How to write a data frame to a File

To save a file with your data simply use the `to_csv` attribute

In [17]:
df.to_csv('myDataFrame.csv')
#df.to_csv('myDataFrame.csv', sep='\t')

# Rename columns

To change the name of a colum use the `rename` attribute

In [18]:
df = df.rename(columns={"Petal_x": "Petalo_x"})

df.head()

Petalo_x  Petal_y  Sepal_x  Sepal_y         Type
0       5.1      3.5      1.4      0.2  Iris-setosa
1       4.9      3.0      1.4      0.2  Iris-setosa
2       4.7      3.2      1.3      0.2  Iris-setosa
3       4.6      3.1      1.5      0.2  Iris-setosa
4       5.0      3.6      1.4      0.2  Iris-setosa

In [19]:
df = df.rename(columns={"Petal_y": "Petalo_y"})

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type
0       5.1       3.5      1.4      0.2  Iris-setosa
1       4.9       3.0      1.4      0.2  Iris-setosa
2       4.7       3.2      1.3      0.2  Iris-setosa
3       4.6       3.1      1.5      0.2  Iris-setosa
4       5.0       3.6      1.4      0.2  Iris-setosa

# Selection of colums

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one column (variable), use one of the following alternatives.

In [20]:
a = df.Petalo_x
b = df["Petalo_x"]
c = df.loc[:, "Petalo_x"]
d = df.iloc[:, 1]

print(d)

#df[["Gender", "GenderGroup"]]

0      3.5
1      3.0
2      3.2
3      3.1
4      3.6
      ... 
145    3.0
146    2.5
147    3.0
148    3.4
149    3.0
Name: Petalo_y, Length: 150, dtype: float64


# Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one observation or for one variable.  There are several ways to do this, as shown below.

Lets say we would like to splice our data frame and select only specific portions of our data.  There are three different ways of doing so.

1. .loc()
2. .iloc()
3. .ix()

We will cover the .loc() and .iloc() splicing functions.


The attibute **.loc()** uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [21]:
# Return all observations of CWDistance
df.loc[:,"Sepal_x"]

# Return a subset of observations of CWDistance
df.loc[:9, "Sepal_x"]

# Select all rows for multiple columns, ["Gender", "GenderGroup"]
df.loc[:,["Sepal_x", "Sepal_y"]]

# Select multiple columns, ["Gender", "GenderGroup"]me
keep = ['Sepal_x', 'Sepal_y']
df_gender = df[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
#df.loc[4:9, ["CWDistance", "Height", "Wingspan"]]

# Select range of rows for all columns
#df.loc[10:15,:]



The attribute **iloc()** is an integer based slicing.

In [22]:
# .
df.iloc[:, :4]

# .
df.iloc[:4, :]

# .
df.iloc[:, 3:7]

# .
df.iloc[4:8, 2:4]

# This is incorrect:
#df.iloc[1:5, ["Gender", "GenderGroup"]]

Sepal_x  Sepal_y
4      1.4      0.2
5      1.7      0.4
6      1.4      0.3
7      1.5      0.2

# Get unique existing values

List unique values in the one of the columns


In [23]:
# List unique values in the df['Gender'] column
df.Sepal_x.unique()

array([1.4, 1.3, 1.5, 1.7, 1.6, 1.1, 1.2, 1. , 1.9, 4.7, 4.5, 4.9, 4. ,
       4.6, 3.3, 3.9, 3.5, 4.2, 3.6, 4.4, 4.1, 4.8, 4.3, 5. , 3.8, 3.7,
       5.1, 3. , 6. , 5.9, 5.6, 5.8, 6.6, 6.3, 6.1, 5.3, 5.5, 6.7, 6.9,
       5.7, 6.4, 5.4, 5.2])

In [24]:
# Lets explore df["GenderGroup] as well
df.Sepal_y.unique()

array([0.2, 0.4, 0.3, 0.1, 0.5, 0.6, 1.4, 1.5, 1.3, 1.6, 1. , 1.1, 1.8,
       1.2, 1.7, 2.5, 1.9, 2.1, 2.2, 2. , 2.4, 2.3])

# Filter, Sort and Groupby



With **Filter** you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

In [25]:
df[df["Petalo_x"] >= 5]

Petalo_x  Petalo_y  Sepal_x  Sepal_y            Type
0         5.1       3.5      1.4      0.2     Iris-setosa
4         5.0       3.6      1.4      0.2     Iris-setosa
5         5.4       3.9      1.7      0.4     Iris-setosa
7         5.0       3.4      1.5      0.2     Iris-setosa
10        5.4       3.7      1.5      0.2     Iris-setosa
..        ...       ...      ...      ...             ...
145       6.7       3.0      5.2      2.3  Iris-virginica
146       6.3       2.5      5.0      1.9  Iris-virginica
147       6.5       3.0      5.2      2.0  Iris-virginica
148       6.2       3.4      5.4      2.3  Iris-virginica
149       5.9       3.0      5.1      1.8  Iris-virginica

[128 rows x 5 columns]

With **Sort** is possible to sort values in a certain column in an ascending order using `df.sort_values("ColumnName")` or in descending order using `df.sort_values(ColumnName, ascending=False)`.

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using `df.sort_values([Column1Name,Column2Name],ascending=[True,False])`


In [26]:
df.sort_values("Petalo_y")
#df.sort_values("Height",ascending=False)

Petalo_x  Petalo_y  Sepal_x  Sepal_y             Type
60        5.0       2.0      3.5      1.0  Iris-versicolor
62        6.0       2.2      4.0      1.0  Iris-versicolor
119       6.0       2.2      5.0      1.5   Iris-virginica
68        6.2       2.2      4.5      1.5  Iris-versicolor
41        4.5       2.3      1.3      0.3      Iris-setosa
..        ...       ...      ...      ...              ...
16        5.4       3.9      1.3      0.4      Iris-setosa
14        5.8       4.0      1.2      0.2      Iris-setosa
32        5.2       4.1      1.5      0.1      Iris-setosa
33        5.5       4.2      1.4      0.2      Iris-setosa
15        5.7       4.4      1.5      0.4      Iris-setosa

[150 rows x 5 columns]

The attribute **Groupby** involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.

In [27]:
df.groupby(['Sepal_x'])


Size of each group

In [28]:
df.groupby(['Sepal_x']).size()

df.groupby(['Sepal_x','Sepal_y']).size()

Sepal_x  Sepal_y
1.0      0.2        1
1.1      0.1        1
1.2      0.2        2
1.3      0.2        4
         0.3        2
                   ..
6.4      2.0        1
6.6      2.1        1
6.7      2.0        1
         2.2        1
6.9      2.3        1
Length: 102, dtype: int64

This output indicates that we have two types of combinations.

* Case 1: Gender = F & Gender Group = 1
* Case 2: Gender = M & GenderGroup = 2.  

This validates our initial assumption that these two fields essentially portray the same information.

# Data Cleaning: handle with missing data

Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  

Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr.

In [29]:
df.isnull()


Petalo_x  Petalo_y  Sepal_x  Sepal_y   Type
0       False     False    False    False  False
1       False     False    False    False  False
2       False     False    False    False  False
3       False     False    False    False  False
4       False     False    False    False  False
..        ...       ...      ...      ...    ...
145     False     False    False    False  False
146     False     False    False    False  False
147     False     False    False    False  False
148     False     False    False    False  False
149     False     False    False    False  False

[150 rows x 5 columns]

Unfortunately, our output indicates that some of our columns contain missing values so we are no able to continue on doing analysis with those colums

In [30]:
df.notnull()

Petalo_x  Petalo_y  Sepal_x  Sepal_y  Type
0        True      True     True     True  True
1        True      True     True     True  True
2        True      True     True     True  True
3        True      True     True     True  True
4        True      True     True     True  True
..        ...       ...      ...      ...   ...
145      True      True     True     True  True
146      True      True     True     True  True
147      True      True     True     True  True
148      True      True     True     True  True
149      True      True     True     True  True

[150 rows x 5 columns]

In [31]:
df.isnull().sum()
df.notnull().sum()

Petalo_x    150
Petalo_y    150
Sepal_x     150
Sepal_y     150
Type        150
dtype: int64

Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

In [32]:
print( df.Type.notnull().sum() )

print( pd.isnull(df.Type).sum() )

150
0


In [33]:
# Extract all non-missing values of one of the columns into a new variable
x = df.Petalo_x.dropna().describe()
x.describe()

count      8.000000
mean      23.271425
std       51.247113
min        0.828066
25%        4.900000
50%        5.821667
75%        6.775000
max      150.000000
Name: Petalo_x, dtype: float64

# Add and eliminate columns

In some cases it is useful to create or eiminate new columns

In [34]:
df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type
0       5.1       3.5      1.4      0.2  Iris-setosa
1       4.9       3.0      1.4      0.2  Iris-setosa
2       4.7       3.2      1.3      0.2  Iris-setosa
3       4.6       3.1      1.5      0.2  Iris-setosa
4       5.0       3.6      1.4      0.2  Iris-setosa

In [35]:
# Add a new column with new data

# Create a column data
NewColumnData = df.Sepal_x/df.Sepal_x

# Insert that column in the data frame
df.insert(5, "ColumnInserted", NewColumnData, True)

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type  ColumnInserted
0       5.1       3.5      1.4      0.2  Iris-setosa             1.0
1       4.9       3.0      1.4      0.2  Iris-setosa             1.0
2       4.7       3.2      1.3      0.2  Iris-setosa             1.0
3       4.6       3.1      1.5      0.2  Iris-setosa             1.0
4       5.0       3.6      1.4      0.2  Iris-setosa             1.0

In [36]:
# Eliminate inserted column
df.drop("ColumnInserted", axis=1, inplace = True)
#df.drop(columns=['ColumnInserted'], inplace = True)
# Remove three columns as index base
#df.drop(df.columns[[12]], axis = 1, inplace = True)

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type
0       5.1       3.5      1.4      0.2  Iris-setosa
1       4.9       3.0      1.4      0.2  Iris-setosa
2       4.7       3.2      1.3      0.2  Iris-setosa
3       4.6       3.1      1.5      0.2  Iris-setosa
4       5.0       3.6      1.4      0.2  Iris-setosa

In [37]:
# Add new column derived from existing columns

# The new column is a function of another column
df["Petalo_xx"] = df["Petalo_x"] * 2

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type  Petalo_xx
0       5.1       3.5      1.4      0.2  Iris-setosa       10.2
1       4.9       3.0      1.4      0.2  Iris-setosa        9.8
2       4.7       3.2      1.3      0.2  Iris-setosa        9.4
3       4.6       3.1      1.5      0.2  Iris-setosa        9.2
4       5.0       3.6      1.4      0.2  Iris-setosa       10.0

In [38]:
# Eliminate inserted column
df.drop("Petalo_xx", axis=1, inplace = True)

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type
0       5.1       3.5      1.4      0.2  Iris-setosa
1       4.9       3.0      1.4      0.2  Iris-setosa
2       4.7       3.2      1.3      0.2  Iris-setosa
3       4.6       3.1      1.5      0.2  Iris-setosa
4       5.0       3.6      1.4      0.2  Iris-setosa

In [39]:
# Add a new column with text labels reflecting the code's meaning

df["Type"] = df.Type.replace({1: "IS", 2: "IVC", 3: "IC"})

# Show the first 5 rows of the created data frame
df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y         Type
0       5.1       3.5      1.4      0.2  Iris-setosa
1       4.9       3.0      1.4      0.2  Iris-setosa
2       4.7       3.2      1.3      0.2  Iris-setosa
3       4.6       3.1      1.5      0.2  Iris-setosa
4       5.0       3.6      1.4      0.2  Iris-setosa

In [40]:
# Eliminate inserted column
df.drop("Type", axis=1, inplace = True)
#df.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y
0       5.1       3.5      1.4      0.2
1       4.9       3.0      1.4      0.2
2       4.7       3.2      1.3      0.2
3       4.6       3.1      1.5      0.2
4       5.0       3.6      1.4      0.2

In [41]:
# Add a new column with strata based on these cut points

# Create a column data
NewColumnData = df.Petalo_x/df.Petalo_x

# Insert that column in the data frame
df.insert(1, "ColumnStrata", NewColumnData, True)

df["ColumnStrata"] = pd.cut(df.Petalo_y, [60., 63., 66., 69., 72., 75., 78.])

# Show the first 5 rows of the created data frame
df.head()

Petalo_x ColumnStrata  Petalo_y  Sepal_x  Sepal_y
0       5.1          NaN       3.5      1.4      0.2
1       4.9          NaN       3.0      1.4      0.2
2       4.7          NaN       3.2      1.3      0.2
3       4.6          NaN       3.1      1.5      0.2
4       5.0          NaN       3.6      1.4      0.2

In [42]:
# Eliminate inserted column
df.drop("ColumnStrata", axis=1, inplace = True)

df.head()

Petalo_x  Petalo_y  Sepal_x  Sepal_y
0       5.1       3.5      1.4      0.2
1       4.9       3.0      1.4      0.2
2       4.7       3.2      1.3      0.2
3       4.6       3.1      1.5      0.2
4       5.0       3.6      1.4      0.2

In [43]:
# Drop several "unused" columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

# Add and eliminate rows

In some cases it is requiered to add new observations (rows) to the data set

In [44]:
df.tail()

Petalo_x  Petalo_y  Sepal_x  Sepal_y
145       6.7       3.0      5.2      2.3
146       6.3       2.5      5.0      1.9
147       6.5       3.0      5.2      2.0
148       6.2       3.4      5.4      2.3
149       5.9       3.0      5.1      1.8

In [45]:
df.loc[len(df.index)] = [26, 24, 'F', 1, 'Y', 1, 66, 'NaN', 68, 'N', 0, 3]

df.tail()


ValueError: cannot set a row with mismatched columns

In [ ]:
# Eliminate inserted row
df.drop([28], inplace = True )

df.tail()


# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [ ]:
# Drop unused columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars]

# Drop rows with any missing values
#df = df.dropna()

# Drop unused columns and drop rows with any missing values
#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars].dropna()

df


# Final remarks


- The understanding of your dataset is essential
    - Number of observations
    - Variables
    - Data types: numerical or categorial
    - What are my variables of interest

- There are several ways to do the same thing

- Cleaning your dataset (dropping out rows with any missing values) is a good practice

- The **Pandas** library provides fancy, high-performance, easy-to-use data structures and data analysis tools


# Activity: work with the iris dataset

Repeat this tutorial with the iris data set and respond to the following inquiries

1. Calculate the statistical summary for each quantitative variables. Explain the results
    - Identify the name of each column:
    'Petal_x', 'Petal_y', 'Sepal_x', 'Sepal_y', 'Type'
    - Identify the type of each column:
    Petal_x    float64
    Petal_y    float64
    Sepal_x    float64
    Sepal_y    float64
    Type        object
    - Minimum, maximum, mean, average, median, standar deviation:
    count    150.000000
    mean       5.843333
    std        0.828066
    min        4.300000
    25%        5.100000
    50%        5.800000
    75%        6.400000
    max        7.900000

2. Are there missing data? If so, create a new dataset containing only the rows with the non-missing data


3. Create a new dataset containing only the petal width and length and the type of Flower


4. Create a new dataset containing only the setal width and length and the type of Flower


5. Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column
